kmod/core/core.c

```diff
@@ -661,31 +682,22 @@ static int kpatch_write_relocations(struct kpatch_module *kpmod,
        unsigned long core = (unsigned long)kpmod->mod->module_core;
        unsigned long core_size = kpmod->mod->core_size;
 #endif
-	unsigned long src;
+	bool vmlinux = !strcmp(object->name, "vmlinux");
 
 	list_for_each_entry(dynrela, &object->dynrelas, list) {
-		if (!strcmp(object->name, "vmlinux")) {
-			ret = kpatch_verify_symbol_match(dynrela->name,
-							 dynrela->src);
-			if (ret)
-				return ret;
-		} else {
-			/* module, dynrela->src needs to be discovered */
-
-			if (dynrela->external)
-				src = kpatch_find_external_symbol(kpmod,
-								  dynrela->name);
-			else
-				src = kpatch_find_module_symbol(object->mod,
-								dynrela->name);
-
-			if (!src) {
-				pr_err("unable to find symbol '%s'\n",
-				       dynrela->name);
-				return -EINVAL;
-			}
-
-			dynrela->src = src;
+		if (dynrela->external)
+			ret = kpatch_find_external_symbol(kpmod->mod->name,
+							  dynrela->name,
+							  dynrela->sympos,
+							  &dynrela->src);
+		else
+			ret = kpatch_find_object_symbol(vmlinux ? NULL : object->name,
```

Comment:
@jpoimboe : I think ```kpatch_find_object_symbol``` can just setup ```args.objname``` accordingly (NULL or objname) and then ```kpatch_kallsyms_callback``` should just work without any changes.  Depends how far down the callchain you want to go.